cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java

```diff
@@ -235,7 +236,8 @@ public void addOutput(String datasetName, Map<String, String> arguments) {
     // compatability for the #setOutput(String, Dataset) method, so delaying the instantiation of this dataset will
     // bring about code complexity without much benefit. Once #setOutput(String, Dataset) is removed, we can postpone
     // this dataset instantiation
-    addOutput(datasetName, new DatasetOutputFormatProvider(datasetName, arguments, getDataset(datasetName, arguments),
+    addOutput(datasetName, new DatasetOutputFormatProvider(datasetName, arguments,
```

Comment:
If the same dataset is used for read, then the access type should be READ_WRITE. That case is not captured.